ingest.py

```diff
@@ -83,60 +84,93 @@ def load(self) -> List[Document]:
 
 def load_single_document(file_path: str) -> Document:
     ext = "." + file_path.rsplit(".", 1)[-1]
-    if ext in LOADER_MAPPING:
-        loader_class, loader_args = LOADER_MAPPING[ext]
-        loader = loader_class(file_path, **loader_args)
-        return loader.load()[0]
 
-    raise ValueError(f"Unsupported file extension '{ext}'")
+    loader_class, loader_args = LOADER_MAPPING[ext]
+    loader = loader_class(file_path, **loader_args)
+    return loader.load()[0]
 
 
-def load_documents(source_dir: str, ignored_files: List[str] = []) -> List[Document]:
+def load_documents(
+    source_dir: str,
+    ingested_files: Optional[List[str]] = None,
+    report_processed_files: bool = False,
+    report_skipped_files: bool = True,
+    report_previously_ingested_files: bool = False,
+    use_progress_bar: bool = True,
+) -> List[Document]:
+
     """
     Loads all documents from the source documents directory, ignoring specified files
     """
-    all_files = []
-    for ext in LOADER_MAPPING:
-        all_files.extend(
-            glob.glob(os.path.join(source_dir, f"**/*{ext}"), recursive=True)
-        )
-    filtered_files = [file_path for file_path in all_files if file_path not in ignored_files]
+    if ingested_files is None:
+        ingested_files : List[str] = []
+
+    filtered_files = []
+    extensions = tuple(LOADER_MAPPING.keys())
+
+    for root, _, files in os.walk(source_dir):
+        for file in files:
+            if file.endswith(extensions):
+                file_path = os.path.join(root, file)
+                if file_path not in ingested_files:
+                    filtered_files.append(file_path)
+                else:
+                    if report_previously_ingested_files:
+                        print(f"Ignored '{file_path}' (already ingested, ignored if updated!)")
+            else:
+                if report_skipped_files:
+                    file_path = os.path.join(root, file)
+                    print(f"Skipping '{file_path}' - unmatched extension")
 
     with Pool(processes=os.cpu_count()) as pool:
         results = []
-        with tqdm(total=len(filtered_files), desc='Loading new documents', ncols=80) as pbar:
+        with (tqdm(total=len(filtered_files), desc='Loading new documents', ncols=80) if use_progress_bar else nullcontext()) as pbar:
             for i, doc in enumerate(pool.imap_unordered(load_single_document, filtered_files)):
                 results.append(doc)
-                pbar.update()
+                if use_progress_bar:
+                    pbar.update()
+                if report_processed_files:
+                    file_path = filtered_files[i]
+                    msg = f"Processed '{file_path}'"
```

Comment:
Due to `pool.imap_unordered()` above, I think the `file_path` processed is not necessarily the one at index `i`.
Probably this is intentional just to print all of them conveniently, just pointing it out, and in case we could change to `imap()`.

Reply:
I made a test case, the order is indeed not guaranteed.

I updated the code so that the paths can be reported as they are treated.